libsel4test/include/sel4test/test.h

```diff
@@ -96,10 +96,16 @@ typedef struct test_type {
     test_result_t (*run_test)(struct testcase *test, uintptr_t e);
 } ALIGN(32) test_type_t;
 
+#if defined(__has_attribute) && __has_attribute(retain)
+#define ATTR_USED_RETAIN __attribute__((used,retain))
+#else
+#define ATTR_USED_RETAIN __attribute__((used))
+#endif
+
 /* Declare a test type.
  * For now, we put the test types in a separate elf section. */
 #define DEFINE_TEST_TYPE(_name, _id, _set_up_test_type, _tear_down_test_type, _set_up, _tear_down, _run_test) \
-    __attribute__((used)) __attribute__((section("_test_type"))) struct test_type TEST_TYPE_ ##_name = { \
+    ATTR_USED_RETAIN __attribute__((section("_test_type"))) struct test_type TEST_TYPE_ ##_name = { \
```

Comment:
Adding `KEEP` to a linker script would make more sense, but I don't think there is a linker script where this could be added.

Could you split the line and put the attributes on their own line please?